nxviz/edges.py

```diff
@@ -82,10 +100,21 @@ def draw(
     to the matplotlib Patch constructor;
     see `nxviz.lines` for more information.
     """
+    nt = node_table(G)
     et = edge_table(G)
     if ax is None:
         ax = plt.gca()
-    edge_color = edge_colors(et, color_by)
+    if color_by in ("source_node_color", "target_node_color"):
+        if not isinstance(G, nx.DiGraph):
+            raise ValueError(
+                "Special values of `color_by`," " can only be set for directed graphs"
+            )
+        elif not node_color_by:
+            raise ValueError(
+                "When setting `color_by` to special values,"
+                "`node_color_by` also needs to be set."
+            )
```

Comment:
Can we defer the validation of arguments to the `edge_colors` function, i.e. one level of abstraction down? I think doing so would help keep this function, `draw`, a bit cleaner.

Reply:
I thought about that, we don't have `G` in `edge_colors` function. passing `G` to that is an option. By passing `G` there is no need for passing `et` and 'nt' because they can be build inside `edge_colors`. What do you think?

Reply:
I see, yes, you're right.

There is a second pattern here that might work a bit better - could you refactor out the validation steps into a separate function and call that function? The intent here is to keep the `draw` function cleaner. A proposed signature could be `validate_color_by(G: nx.DiGraph, color_by: Hashable, node_color_by: Hashable) -> None`.

Reply:
Yes, I will try this.